node/src/accountant/payment_adjuster/mod.rs

```diff
@@ -0,0 +1,2340 @@
+        )?;
+
+        if zero_affordable_accounts_found(&processed_accounts) {
+            return Err(PaymentAdjusterError::AllAccountsEliminated);
+        }
+
+        match processed_accounts {
+            Either::Left(non_exhausted_accounts) => {
+                let original_cw_service_fee_balance_minor =
+                    self.inner.original_cw_service_fee_balance_minor();
+                let exhaustive_affordable_accounts = exhaust_cw_balance_entirely(
+                    non_exhausted_accounts,
+                    original_cw_service_fee_balance_minor,
+                );
+                Ok(exhaustive_affordable_accounts)
+            }
+            Either::Right(finalized_accounts) => Ok(finalized_accounts),
+        }
+    }
+
+    fn propose_adjustments_recursively<AR, RT>(
+        &mut self,
+        unresolved_accounts: Vec<WeightedPayable>,
+        adjustment_runner: AR,
+    ) -> RT
+    where
+        AR: AdjustmentRunner<ReturnType = RT>,
+    {
+        diagnostics!(
+            "\nUNRESOLVED QUALIFIED ACCOUNTS IN CURRENT ITERATION:",
+            &unresolved_accounts
+        );
+
+        adjustment_runner.adjust_accounts(self, unresolved_accounts)
+    }
+
+    fn begin_with_adjustment_by_transaction_fee(
+        &mut self,
+        weighted_accounts: Vec<WeightedPayable>,
+        already_known_affordable_transaction_count: u16,
+    ) -> Result<
+        Either<Vec<AdjustedAccountBeforeFinalization>, Vec<PayableAccount>>,
+        PaymentAdjusterError,
+    > {
+        let error_factory = LateServiceFeeSingleTxErrorFactory::new(&weighted_accounts);
+
+        let weighted_accounts_affordable_by_transaction_fee =
+            dump_unaffordable_accounts_by_transaction_fee(
+                weighted_accounts,
+                already_known_affordable_transaction_count,
+            );
+
+        let cw_service_fee_balance_minor = self.inner.original_cw_service_fee_balance_minor();
+
+        if self.analyzer.recheck_if_service_fee_adjustment_is_needed(
+            &weighted_accounts_affordable_by_transaction_fee,
+            cw_service_fee_balance_minor,
+            error_factory,
+            &self.logger,
+        )? {
+            diagnostics!("STILL NECESSARY TO CONTINUE BY ADJUSTMENT IN BALANCES");
+
+            let adjustment_result_before_verification = self
+                .propose_possible_adjustment_recursively(
+                    weighted_accounts_affordable_by_transaction_fee,
+                );
+
+            Ok(Either::Left(adjustment_result_before_verification))
+        } else {
+            let accounts_not_needing_adjustment =
+                convert_collection(weighted_accounts_affordable_by_transaction_fee);
+
+            Ok(Either::Right(accounts_not_needing_adjustment))
+        }
+    }
+
+    fn propose_possible_adjustment_recursively(
+        &mut self,
+        weighed_accounts: Vec<WeightedPayable>,
+    ) -> Vec<AdjustedAccountBeforeFinalization> {
```

Comment:
This is a vague comment. I apologise for that.

It's here because I feel the need to refactor this function such that the recursive calls are visible inside the scope of the function.

If the function follows this criteria:

```rust
fn recursive_fn() -> Something {
   if breaking_condition {
       return Something;
   } else {
       recursive_fn()
   }
}
```

Reply:
A bit of refactoring got us here, I hope you'll find it useful during your changes:

```rust
    fn propose_possible_adjustment_recursively_new(
        &mut self,
        weighed_accounts: Vec<WeightedPayable>,
    ) -> Vec<AdjustedAccountBeforeFinalization> {
        let disqualification_arbiter = &self.disqualification_arbiter;
        let unallocated_cw_service_fee_balance =
            self.inner.unallocated_cw_service_fee_balance_minor();
        let logger = &self.logger;

        let current_iteration_result = self.service_fee_adjuster.perform_adjustment_by_service_fee(
            weighed_accounts,
            disqualification_arbiter,
            unallocated_cw_service_fee_balance,
            logger,
        );

        let remaining_undecided_accounts = current_iteration_result.remaining_undecided_accounts;
        let decided_accounts: Vec<AdjustedAccountBeforeFinalization> =
            current_iteration_result.decided_accounts.into();
        if remaining_undecided_accounts.is_empty() {
            return decided_accounts;
        }

        if !decided_accounts.is_empty() {
            self.adjust_remaining_unallocated_cw_balance_down(&decided_accounts)
        }

        let merged = if self.is_cw_balance_under_limit(&remaining_undecided_accounts) {
            // Fast return after a direct conversion into the expected type
            Self::add_decided_accounts(
                convert_collection(remaining_undecided_accounts),
                decided_accounts,
            )
        } else {
            Self::add_decided_accounts(
                self.propose_possible_adjustment_recursively(remaining_undecided_accounts),
                decided_accounts,
            )
        };

        diagnostics!(
            "\nFINAL SET OF ADJUSTED ACCOUNTS IN CURRENT ITERATION:",
            &merged
        );

        merged
    }
 ```